### ISSUE TITLE
Quest - Add UI for Tasks

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#844]
**Name of feature to enhance**
Add UI to be able to view Tasks 

**Description of feature**
Add UI with Tasks. Labels can be either text or icons 

**Describe the enhancement**
See screenshots 

<img width="662" alt="Screen Shot 2021-12-07 at 12 23 42 PM" src="https://user-images.githubusercontent.com/4540684/145008666-ccd9246f-e335-452a-aef4-bdd487f4b308.png">

Link to icons  
1. Overdue - https://drive.google.com/file/d/10RZSKD-JAgsbSoNcfVZiDYGQplz38BF8/view?usp=sharing
2. Due - https://drive.google.com/file/d/1Y5OiTYwkle5Izk_Bn58QMhyfU8gZ7UvS/view?usp=sharing

**Additional context** 
Add support for both labels and Icons  - controlled vis configuration settings we can support both and other alternatives

**Pre-Req task** 
- #857

**Comment by pld** [comment]
This is kinda-dependent on Task generation working, but we can get around that by manually adding some tasks to the FHIR API for now

Are we syncing Tasks? We'll need to enable that if not.

@f-odhiambo what's the different btwn the left and right screens above? in what case do you use one UI or the other, is that based on a configuration setting?

should this show all tasks that have been synced to the device? or only those tasks for the logged in user?

**Comment by pld** [comment]
Also, I would classify this as ANC, I think this is a core/engine feature we'll want in Quest.

The example here is the generic feature with an ANC configuration.

**Comment by pld** [comment]
cool  thanks for

(1) we'll probably want to push that to managed by configurations so it can vary for different projects
(2) if I'm a supervisor, and I want to see all the tasks for the HCWs I manage, would I use this same view or is that a separate list? I suppose, regardless we'd want to re-use the same code just reconfigured, therefore we should architect to

a) handle syncing tasks for a set of HCWs, not a individual HCW
b) be able to configure the UI to show the HCW the task is assigned to

**Comment by f-odhiambo** [comment]
@rowo @HenryRae Kindly advise on whether to go with test labels or go with icons

**Comment by pld** [comment]
Well, I don't think it should matter, if we control it via configuration settings we can support both and other alternatives

**Comment by f-odhiambo** [comment]
We can use JetPack compose for this 
CC @qiarie

**Comment by FikriMilano** [comment]
Okay, so in conclusion this feature will be added in engine module and labels/icons will be controlled by the configuration.
@f-odhiambo do you know where I can get any sample or json for the Tasks?

**Comment by f-odhiambo** [comment]
> Okay, so in conclusion this feature will be added in engine module and labels/icons will be controlled by the configuration.
> @f-odhiambo do you know where I can get any sample or json for the Tasks?

I will add those before end of day today

**Comment by FikriMilano** [comment]
@f-odhiambo 
I have several questions:
1. Where supposed the "Village" data came from? Is it district (which part of the Patient's address)? or was that just a placeholder?
2. What is "#[ID]"? Is it taskId or patientId?
3. What happens when there's more than one CarePlan for a single Patient? How do we know which Task points to which CarePlan to know if a Task is overdue or not? Or was a Patient supposed to only have one CarePlan?

**Comment by f-odhiambo** [comment]
**Some notes here**
1. A CHW (Practioner) will be assigned Tasks to conduct based on the location/ facility they are assigned to
2. The  CHW also below to a CareTeam that are assigned to patients to be able to make health follow-up for their assigned patients 
3. All Tasks will be synced based on the ID of the practitioner who is logged in to the application 
4. Each task will have a status as defined [here](https://www.hl7.org/fhir/valueset-task-status.html#expansion) but at the moment  for this we can use either requested (upcoming), ready (due/overdue) & completed (done/fulfilled)

More about Tasing can be found [here](https://smartregister.atlassian.net/wiki/spaces/Documentation/pages/1156251738/FHIR+Task+Based+Referrals+Software+Requirements+Specification) 

**UI/UX**
<img src="https://user-images.githubusercontent.com/4540684/149730602-9b6b8540-b6a4-41f7-8200-bb939f9b6a22.png" width="200" height="400" />

Based on the above mock 

1. Where supposed the "Village" data came from? Is it the Patient's district? or was that just a placeholder? Answered: placeholder for now. - _**This will be the location name of the current user based on the address that  was added during registration see mock above, should be similar to ANC patient listing/register**_
2. What is "#[ID]"? Is it taskId or patientId? **_This will be the Task ID_** 
3. What happens when there's more than one CarePlan for a single Patient? How do we know which Task points to which CarePlan to know if a Task is overdue or not? Or was a Patient supposed to only have one CarePlan?  The presumption here is to query and filter all the Tasks that are assigned to the practitioner. This is irrespective of whether a client has 1 or multiple CarePalns assigned to them thus this may have the following queries 
- Practioner Details  - http://turn-fhir.smartregister.org/fhir/Practitioner/6744 - retrieved post login 
- Practioner tasks - http://turn-fhir.smartregister.org/fhir/Task/?owner=6744 - Query all tasks assigned to a practitioner 
4. Where can we get the data for distance? e.g 2 km - **_This was supposed to show the approximate distance from the Practioner's location to the patients/family residence location  - To be discussed /implementation later_** 
5. What indicators will be used to either display the distance (e.g 2 km) or action (e.g Call)

**Comment by f-odhiambo** [comment]
The practitioner details will be done via this issue  - #857

**Comment by FikriMilano** [comment]
Thanks a lot for that feedback.

Regarding question number 3:
So how do we know if a task is overdue or not? From what I understand now, it should be at least had the status as "ready" and there should be something more, e.g the due Date. I thought we can know if a Task is overdue or not by it's CarePlan period.
https://turn-fhir.smartregister.org/fhir/CarePlan/ee802d1f-3168-4598-a113-8a69f268e418

**Comment by f-odhiambo** [comment]
The CarePlan period may be prep planned and have period dates but can be updated based on the task scheduled for the client. For Tasks, you can query and check the status of the task as follows for this 1st implementation  
1. Requested  - To Do, planned 
2. Ready  - Due or overdue 
3. Completed  - DONE

**Comment by FikriMilano** [comment]
There is a period element in Task which I think we can use for this. By comparing the current date with end date, we can know it's overdue date.
https://www.hl7.org/fhir/task.html#resource
https://www.hl7.org/fhir/task-definitions.html#Task.restriction.period

**Comment by f-odhiambo** [comment]
> There is a period element in Task which I think we can use for this. By comparing the current date with end date, we can know it's overdue date.
> https://www.hl7.org/fhir/task.html#resource
> https://www.hl7.org/fhir/task-definitions.html#Task.restriction.period

Sure

**Comment by FikriMilano** [comment]
Thanks, could you also update the Task samples with period too?
http://turn-fhir.smartregister.org/fhir/Task/?owner=6744

**Comment by f-odhiambo** [comment]
Done, I think the filter should check both status and date ..using the below fields to track this 

```
    "executionPeriod": {
    "start": "2021-10-01T08:45:05+10:00"
    },"restriction": {
    "repetitions": 1,
        "period": {
        "end": "2021-10-02T09:45:05+10:00"
        }
    }
```

**Comment by FikriMilano** [comment]
> Done, I think the filter should check both status and date ..using the below fields to track this
Sure, just to make sure, so we are filtering the Tasks that only has Requested/Ready/Completed by status?

**Comment by f-odhiambo** [comment]
For this POC we can do that and then later expand the filters

**Comment by FikriMilano** [comment]
Okay, last one, should we display the same Blue label for Requested and Completed? or do you have other plans?

**Comment by f-odhiambo** [comment]
> Okay, last one, should we display the same Blue label for Requested and Completed? or do you have other plans?

Requested - To Do, planned - Blue
Ready - Due or overdue - Red
Completed - DONE - Green

**Comment by f-odhiambo** [comment]
Test data/patient https://turn-fhir.smartregister.org/fhir/Patient/6741/$everything

**Comment by pld** [comment]
Updating the title to be about Quest, it's important this is all in Quest and Engine, there should be nothing in ANC or any use-case specific apps.

If there's anything that can't be done generic please schedule a meeting to discuss, that will require re-architecting before we proceed, thanks

**Comment by FikriMilano** [comment]
> Updating the title to be about Quest, it's important this is all in Quest and Engine, there should be nothing in ANC or any use-case specific apps.
> 
> If there's anything that can't be done generic please schedule a meeting to discuss, that will require re-architecting before we proceed, thanks

Okay, let me try to find a generic solution for this.

**Comment by pld** [comment]
I'm not sure of the sizing, but I'd guess this is a medium, not extra small, mainly this UI will require configuration code to keep it generic.

Additionally two questions:

1) how can we re-use the existing list views? I'm not familiar w/the current list rendering, the ideal here is to have a list component that can render any FHIR resources and here we are writing a concrete class, sub-classing, or otherwise reusing that general list renderer. In an ideal configurable state, we could load another FHIR file that defines the configuration for this resource-list-view, but that can be for future work if we do not do that for other views yet.
2) how do you get to the tasking list? is this in a menu, or navigation somewhere, can add a description of the UX for that and a screenshot please

**Comment by FikriMilano** [comment]
1. @pld That's interesting, what I have in mind is to create some kind of ListFragment in Engine module then override some method from that class through the AndroidApp modules (ANC, Quest) that could control what view item would be rendered and onClick handlers. The view item could be any FHIR resource.
From what I know, at the moment there's only Task resource that needs to be rendered as a generic list.
2. I retrieve the tasking list from HAPI and load the Patients from DB based on the Task owner (HCW) then map both to a single model so that it could carry the necessary data. Here's the tasking list source https://turn-fhir.smartregister.org/fhir/Task/?owner=6744.
For the ANC case, the tasking list could be accessed from the Bottom Navigation Bar.

The current implementation at the moment is just concrete class in ANC, but I will move this implementation to Engine and make it generic. Will have a further discussion with @ellykits regarding this and the configuration part.

cc @f-odhiambo

Screenshot (ANC):
<img src="https://user-images.githubusercontent.com/62053304/150324375-0078b782-cb05-4376-8709-65135cc02629.png" width="256"/>

**Comment by pld** [comment]
That's seems reasonable, what about for the patient list view in Quest? can we reuse that?

Also, note there's no need to do any integration w/ANC for this, we'll be deprecating that module as much as we can as long as we're on track to get R2 out on Quest CC @dubdabasoduba

**Comment by ellykits** [comment]
From our discussion with @FikriMilano we agreed on the following:

- Implement a task register extending the [ComposeRegisterFragment](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/register/ComposeRegisterFragment.kt) abstract class which provides generic functionalities for `handling click listeners on list rows` and also provides a `paginated` view accepting a list
- Above can be implemented as an abstract `BaseTaskRegister<T>` in engine module where `T` is the type of the input model (Can be any Resource/List of resources). This should allow for implementing multiple Task registers in the application; for a single task register this can be a concrete class.
- Implement a`LazyColumn` composable view like the one used in the existing apps, use it in the Task register fragment.
- Reuse [PatientRegisterRowViewConfiguration](https://github.com/opensrp/fhircore/blob/main/android/quest/src/main/java/org/smartregister/fhircore/quest/configuration/view/PatientRegisterRowViewConfiguration.kt) to configure row items of the list. We will need to move this model to engine module and rename it appropriately to a more generic name like `RegisterRowViewConfiguration`

**Comment by FikriMilano** [comment]
@f-odhiambo I need some help with the icons, there are cuts at the left and bottom side of it.
Happens with the due and overdue icons.

SVG (From the source above)
![ic_due_svg](https://user-images.githubusercontent.com/62053304/152285220-056ff0ef-6f10-4ca1-9ee5-70bd974b2ce7.PNG)

XML (Converted in Android Studio)
![ic_due](https://user-images.githubusercontent.com/62053304/152285257-1538cb96-e751-413f-bd9e-a22731a614ce.PNG)

**Comment by f-odhiambo** [comment]
@rowo @HenryRae Kindly assist  in getting an icon set in SCG for this ticket for screen two in the main issue description . Thanks

**Comment by FikriMilano** [comment]
Kindly also provide this icon
@rowo @HenryRae @f-odhiambo 

<img width="256" alt="icon" src="https://user-images.githubusercontent.com/62053304/153136493-c85dcaea-8262-4254-85a8-bbb84b12aaaa.png">

**Comment by HenryRae** [comment]
Please find the attached svgs

![Frame 536](https://user-images.githubusercontent.com/31955469/153140176-5a6faa3e-7e23-4f65-abc9-2c7a98e8fc72.svg)
![x](https://user-images.githubusercontent.com/31955469/153140185-195d2deb-232d-45cd-acb8-610e43589bd0.svg)

**Comment by FikriMilano** [comment]
> Please find the attached svgs
> 
> ![Frame 536](https://user-images.githubusercontent.com/31955469/153140176-5a6faa3e-7e23-4f65-abc9-2c7a98e8fc72.svg) ![x](https://user-images.githubusercontent.com/31955469/153140185-195d2deb-232d-45cd-acb8-610e43589bd0.svg)

Thanks, could you also provide the red overdue too?

**Comment by HenryRae** [comment]
Please find the attached
![Frame 536 - red](https://user-images.githubusercontent.com/31955469/153155142-7b32865d-d9cb-45b1-a5e8-acb4c9319a9c.svg)

**Comment by FikriMilano** [comment]
hi @pld - It's doing good.
So for tasking list, I made some changes on the abstraction and used the same parent class (or view) that is used by patient list then added it's implementation to quest.

I just need to add some test case on the PR and update the icons then it's good to review.
#1023
